### ISSUE TITLE
Error response from daemon: Server error from portlayer: --name: no such executable in PATH.**

### ISSUE THREAD

**Issue by mobla** [vmware/vic#3114]
Could be related to
https://github.com/vmware/vic/issues/2273
https://github.com/vmware/vic/issues/1246
??

VIC build 0.7.0-6507
vic-machine-linux version v0.7.0-6507-6c4627d

Will provide the setup on a need basis. 
Not able to connect to https://wdc-qead-dhcp-10-139-23-239.eng.vmware.com:2378 and collect logs

Steps:
1. Create VCH with TLS

2. docker login to Harbor private registry (https) public CA cert and push images (vmware/powerclicore and busybox)
> root@photon-lJUA39j42 [ ~ ]# docker tag busybox:latest xxxx.eng.vmware.com/mobla-project/busybox:1.0
> root@photon-lJUA39j42 [ ~ ]# docker push xxxx.eng.vmware.com/mobla-project/busybox:1.0
The push refers to a repository [xxxx.eng.vmware.com/mobla-project/busybox]
e88b3f82283b: Pushed

3. docker run via VCH (mistakenly gave -n flag)
> root@photon-lJUA39j42 [ ~ ]# DOCKER_CERT_PATH=/root/vic/vic/VCHTLS1 DOCKER_TLS_VERIFY=1 docker -H xxxx.eng.vmware.com:2376 run -it xxxx.eng.vmware.com/mobla-project/busybox:1.0 **-n hungry-ramanujan**
Unable to find image 'xxxx.eng.vmware.com/mobla-project/busybox:1.0' locally
Pulling from mobla-project/busybox
56bec22e3559: Pull complete
a3ed95caeb02: Pull complete
Digest: sha256:1f092a823e2607ac40df7df7c6725c33c3dada9092120ba719eadd94cd46012b
Status: Downloaded newer image for mobla-project/busybox:1.0
**docker: Error response from daemon: Server error from portlayer: -n: no such executable in PATH.**

docker run via VCH (with --name flag)
> root@photon-lJUA39j42 [ ~ ]# DOCKER_CERT_PATH=/root/vic/vic/VCHTLS1 DOCKER_TLS_VERIFY=1 docker -H xxx.eng.vmware.com:2376 run -it xxxx.eng.vmware.com/mobla-project/busybox:1.0 -**-name hungry-ramanujan**
**docker: Error response from daemon: Server error from portlayer: --name: no such executable in PATH.**

docker run via VCH (with no name flag but still fails)
>root@photon-lJUA39j42 [ ~ ]# DOCKER_CERT_PATH=/root/vic/vic/VCHTLS1 DOCKER_TLS_VERIFY=1 docker -H xxx.eng.vmware.com:2376 run -it xxxx.eng.vmware.com/mobla-project/busybox:1.0
**docker: Error response from daemon: Server error from portlayer: --name: no such executable in PATH.**

Expected behavior: vic-engine must pull the image xxxx.eng.vmware.com/mobla-project/busybox:1.0 from harbor

**Comment by mlh78750** [comment]
@mobla the --name option must be before the image name. All command line stuff after the container image is passed to the container as an alternative ENTRYPOINT (basically the command you want to run in that image).

That being said, after it initially exited, I would think the run with no --name flag as all should have worked. We can track why in this issue. But the -n and --name at the end of the command line resulted in the expected behavior. Note that on the first try you actually did pull the image from harbor.

**Comment by mlh78750** [comment]
@mobla what is the result of ```docker images -a```

**Comment by mobla** [comment]
> root@photon-lJUA39j42 [ ~/vic/vic ]# DOCKER_CERT_PATH=/root/vic/vic/VCHTLS1 DOCKER_TLS_VERIFY=1 docker -H wdc-qead-dhcp-10-139-23-239.eng.vmware.com:2376 images -a
REPOSITORY                                         TAG                 IMAGE ID            CREATED             SIZE
reg-bj.eng.vmware.com/mobla-project/powerclicore   1.0                 e3b924cb9ae8        47 hours ago        419.3 MB
reg-bj.eng.vmware.com/mobla-project/busybox        1.0                 e292aa76ad3b        4 weeks ago         1.093 MB

**Comment by mlh78750** [comment]
and if you run DOCKER_CERT_PATH=/root/vic/vic/VCHTLS1 DOCKER_TLS_VERIFY=1 docker -H xxx.eng.vmware.com:2376 run -it xxxx.eng.vmware.com/mobla-project/busybox:1.0 now it will still say something about --name? 

cc: @fdawg4l @jzt guys, I know that the first few commands should have failed like they did, but why could the last one be failing? are we caching the entrypoint that is provided on the command line with a run command in the image metadata?

**Comment by fdawg4l** [comment]
I can tell you the portlayer storage piece of the puzzle is immutable.  As far as what data is being persisted, imagec is the thing providing the metadata to persist.  So @jzt will definitely be the authority here as to what's going on.

**Comment by mobla** [comment]
Any image is failing to run now, same error! Got stuck
> root@photon-lJUA39j42 [ ~/vic/vic ]# DOCKER_CERT_PATH=/root/vic/vic/VCHTLS1 DOCKER_TLS_VERIFY=1 docker -H xxxxx.eng.vmware.com:2376 run -it xxxx/mobla-project/busybox:1.0
docker: Error response from daemon: Server error from portlayer: --name: no such executable in PATH.

**Comment by mobla** [comment]
logs uploaded
[name-error-bug.tar.gz](https://github.com/vmware/vic/files/586981/name-error-bug.tar.gz)

**Comment by jzt** [comment]
I created a VC cluster using the same VC & ESX builds as @mobla and was able to reproduce this. I've narrowed it down to the container create step altering the in-memory image cache. I suspect there is a pointer nested somewhere in the image metadata struct that is being changed somewhere where it should be a copy. I believe this only manifests if the image is tagged to an external registry, but I can't say why that might be. I can reproduce it every time if I pull the Harbor-tagged busybox before the Docker Hub busybox, but if I pull in the opposite order, the problem goes away. My feeling is that the tag and how the cache uses the tag to look up image metadata might have something to do with it.

### Pulling the Harbor-tagged busybox first:
```
$ docker -H 10.192.101.91:2375 pull sc2-rdops-vm03-dhcp-113-126.eng.vmware.com/zach-project/busybox:1.0
Pulling from zach-project/busybox
56bec22e3559: Pull complete
a3ed95caeb02: Pull complete
Digest: sha256:d8a3a03fdec384e3f637e79747e91cebd921ea3bd2334df5f5bd06c46100ce0e
Status: Downloaded newer image for zach-project/busybox:1.0
```
Pull busybox from Docker Hub:
```
$ docker -H 10.192.101.91:2375 pull busybox
Using default tag: latest
Pulling from library/busybox
56bec22e3559: Already exists
a3ed95caeb02: Already exists
Digest: sha256:98a0bd48d22ff96ca23bfda2fe1cf72034ea803bd79e64a5a5f274aca0f9c51c
Status: Image is up to date for library/busybox:latest
```
Create  a `/bin/date` container with the Harbor busybox image:
```
$ docker -H 10.192.101.91:2375 create sc2-rdops-vm03-dhcp-113-126.eng.vmware.com/zach-project/busybox:1.0 /bin/date
d14ffa41d33fab7164f4054505c981605bf6080e87c20e360cc8dec1df864b45
```
Logged image metadata used by this container (notice the `Cmd:strslice.StrSlice{\"sh\"}` from the fresh image metadata):
```
time="2016-11-13T03:57:21Z" level=info msg="Image config: &container.Config{Hostname:\"4a74292706a0\", Domainname:\"\", User:\"\",
  AttachStdin:false, AttachStdout:false, AttachStderr:false, ExposedPorts:map[nat.Port]struct {}(nil), PublishService:\"\", Tty:false, OpenStdin:false,
  StdinOnce:false, Env:[]string{\"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin\"}, Cmd:strslice.StrSlice{\"sh\"}, ArgsEscaped:false, 
  Image:\"sha256:1679bae2167496818312013654f5c66a16e185d0a0f6b762b53c8558014457c6\", Volumes:map[string]struct {}(nil), WorkingDir:\"\",
  Entrypoint:strslice.StrSlice(nil), NetworkDisabled:false, MacAddress:\"\", OnBuild:[]string(nil), Labels:map[string]string{}, StopSignal:\"\"}"
```
Create a default container with no command:
```
$ docker -H 10.192.101.91:2375 create sc2-rdops-vm03-dhcp-113-126.eng.vmware.com/zach-project/busybox:1.0
7f75e053e2047509fbfd9337a5a444584e0065ded46bcc2849f2ffaf7feeef72
```
Logged image metadata used by this container (notice the `Cmd` stuck on `/bin/date`):
```
time="2016-11-13T03:59:11Z" level=info msg="Image config: &container.Config{Hostname:\"4a74292706a0\", Domainname:\"\", User:\"\",
  AttachStdin:false, AttachStdout:true, AttachStderr:true, ExposedPorts:map[nat.Port]struct {}(nil), PublishService:\"\", Tty:false, OpenStdin:false,
  StdinOnce:false, Env:[]string{\"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin\"}, Cmd:strslice.StrSlice{\"/bin/date\"}, ArgsEscaped:false, 
  Image:\"sha256:1679bae2167496818312013654f5c66a16e185d0a0f6b762b53c8558014457c6\", Volumes:map[string]struct {}(nil), WorkingDir:\"\",
  Entrypoint:strslice.StrSlice(nil), NetworkDisabled:false, MacAddress:\"\", OnBuild:[]string(nil), Labels:map[string]string{}, StopSignal:\"\"}"
```

### Pulling Docker Hub's busybox first (Note: :
```
$ docker -H 10.192.122.233:2375 pull busybox
Using default tag: latest
Pulling from library/busybox
56bec22e3559: Pull complete
a3ed95caeb02: Pull complete
Digest: sha256:98a0bd48d22ff96ca23bfda2fe1cf72034ea803bd79e64a5a5f274aca0f9c51c
Status: Downloaded newer image for library/busybox:latest
```
Pull the Harbor-tagged image next:
```
$ docker -H 10.192.122.233:2375 pull sc2-rdops-vm03-dhcp-113-126.eng.vmware.com/zach-project/busybox:1.0
Pulling from zach-project/busybox
56bec22e3559: Already exists
a3ed95caeb02: Already exists
Digest: sha256:d8a3a03fdec384e3f637e79747e91cebd921ea3bd2334df5f5bd06c46100ce0e
Status: Image is up to date for zach-project/busybox:1.0
```
Create a `/bin/date` container:
```
$ docker -H 10.192.122.233:2375 create -it sc2-rdops-vm03-dhcp-113-126.eng.vmware.com/zach-project/busybox:1.0 /bin/date
e041a18e42781b31b98c401121f85d5d1d3fb98fdde7a95b4549649058fb7503
```
Logs:
```
time="2016-11-13T03:42:06Z" level=info msg="Image config: &container.Config{Hostname:\"4a74292706a0\", Domainname:\"\", User:\"\",
  AttachStdin:false, AttachStdout:false, AttachStderr:false, ExposedPorts:map[nat.Port]struct {}(nil), PublishService:\"\", Tty:false, OpenStdin:false,
  StdinOnce:false, Env:[]string{\"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin\"}, Cmd:strslice.StrSlice{\"sh\"}, ArgsEscaped:false, 
  Image:\"sha256:1679bae2167496818312013654f5c66a16e185d0a0f6b762b53c8558014457c6\", Volumes:map[string]struct {}(nil), WorkingDir:\"\",
  Entrypoint:strslice.StrSlice(nil), NetworkDisabled:false, MacAddress:\"\", OnBuild:[]string(nil), Labels:map[string]string{}, StopSignal:\"\"}"
```
Create a default container with no command:
```
$ docker -H 10.192.122.233:2375 create -it sc2-rdops-vm03-dhcp-113-126.eng.vmware.com/zach-project/busybox:1.0
b0a2c25335b7fa245cc4efad4fa58b1fc956c5ef86f71c67800df5e3d9978e77
```
Logs (note the image metadata `Cmd` is preserved:
```
time="2016-11-13T03:42:16Z" level=info msg="Image config: &container.Config{Hostname:\"4a74292706a0\", Domainname:\"\", User:\"\",
  AttachStdin:false, AttachStdout:false, AttachStderr:false, ExposedPorts:map[nat.Port]struct {}(nil), PublishService:\"\", Tty:false, OpenStdin:false,
  StdinOnce:false, Env:[]string{\"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin\"}, Cmd:strslice.StrSlice{\"sh\"}, ArgsEscaped:false, 
  Image:\"sha256:1679bae2167496818312013654f5c66a16e185d0a0f6b762b53c8558014457c6\", Volumes:map[string]struct {}(nil), WorkingDir:\"\",
  Entrypoint:strslice.StrSlice(nil), NetworkDisabled:false, MacAddress:\"\", OnBuild:[]string(nil), Labels:map[string]string{}, StopSignal:\"\"}"
```

**Comment by jzt** [comment]
Fix here:  #3159

**Comment by mdubya66** [comment]
Fixed in build 6879
